core/jvm/src/unsafe/UnsafeBufferOperationsJvm.kt

```diff
@@ -0,0 +1,158 @@
+ * the best effort basis, meaning that there are no strong zero-copy guarantees
+ * and the copy will be created if it could not be omitted.
+ *
+ * @param buffer a buffer to read from
+ * @param readAction an action that will be invoked on a [ByteBuffer] containing data from [buffer]'s head
+ *
+ * @throws IllegalArgumentException when the [buffer] is empty.
+ *
+ * @sample kotlinx.io.samples.unsafe.UnsafeReadWriteSamplesJvm.writeToByteChannel
+ */
+@UnsafeIoApi
+public inline fun UnsafeBufferOperations.readFromHead(buffer: Buffer, readAction: (ByteBuffer) -> Unit) {
+    readFromHead(buffer) { rawData, pos, limit ->
+        val bb = ByteBuffer.wrap(rawData, pos, limit - pos).slice().asReadOnlyBuffer()
+        readAction(bb)
+        bb.position()
+    }
+}
+
+/**
+ * Provides write access to the buffer, allowing to write data
+ * into a not yet committed portion of the buffer's tail using a [writeAction].
+ *
+ * The [writeAction] receives a [ByteBuffer] representing uncommitted portion of [buffer]'s tail
+ *
+ * It's guaranteed that the size of the [ByteBuffer] is at least [minimumCapacity],
+ * but if the [minimumCapacity] bytes could not be provided for writing,
+ * the method will throw [IllegalStateException].
+ * It is safe to use any [minimumCapacity] value below [maxSafeWriteCapacity], but unless exact minimum number of
+ * available bytes is required, it's recommended to use `1` as [minimumCapacity] value.
+ *
+ * After exiting [writeAction], bytes written to the [ByteBuffer] will be committed to the buffer.
+ * The number of bytes written is determined as a difference between [ByteBuffer.capacity] and [ByteBuffer.remaining].
+ *
+ * If [writeAction] ends execution by throwing an exception, no data will be written to the buffer.
+ *
+ * The data array is passed to the [writeAction] directly from the buffer's internal storage without copying
+ * on the best-effort basis, meaning that there are no strong zero-copy guarantees
+ * and the copy will be created if it could not be omitted.
+ *
+ * @param buffer a buffer to read from
+ * @param minimumCapacity the minimum amount of writable space
+ * @param writeAction an action that will be invoked on a [ByteBuffer] that will be added to a [buffer] by the end of
+ * the call
+ *
+ * @throws IllegalStateException when [minimumCapacity] is too large and could not be fulfilled.
+ *
+ * @sample kotlinx.io.samples.unsafe.UnsafeReadWriteSamplesJvm.readFromByteChannel
+ */
+@UnsafeIoApi
+public inline fun UnsafeBufferOperations.writeToTail(
+    buffer: Buffer,
+    minimumCapacity: Int,
+    writeAction: (ByteBuffer) -> Unit
+) {
+    writeToTail(buffer, minimumCapacity) { rawData, pos, limit ->
+        val bb = ByteBuffer.wrap(rawData, pos, limit - pos).slice()
+        writeAction(bb)
+        bb.position()
+    }
+}
+
+/**
+ * Provides read-only access to [buffer]'s data by filling provided [iovec] array with [ByteBuffer]'s representing
```

Comment:
It is not entirely clear from the documentation what the size of each ByteBuffer in the array will be and, in general, on the basis of which the total number of buffers is determined, is it predictable or not, which array size is better to use

Reply:
The documentation is indeed require some improvements.

However, the API itself was designed with the idea that `iovec`-array will be allocated once and cached by the caller (for instance, as a thread-local variable) to reduce excessive allocations on the write path.
The number of ByteBuffers to be created corresponds to the number of segments a buffer has. Subsequent patches provides an API allowing to find that value, but in general, it's hard to give an advice on the array size (for example, a buffer holding 8000 bytes may have either 1 segment, or multiple).